File: Scheduling/sjf.py
import pandas as pd
import operator

class Process:
    def __init__(self):
        self.pid=input("Enter Process ID : ")
        self.bt=int(input("Enter Burst time : "))
    def get_wt(self,x):
        self.wt=x
    def get_tat(self,x):
        self.tat=x

PID=[]
BT=[]
WT=[]
TAT=[]
n = int(input("No:of processes ? "))
p = []
for i in range(0,n):
    p.append(Process())
p.sort(key=operator.attrgetter('bt'))
for i in range(0, n):
    if i==0:
        p[i].get_tat(p[i].bt)
    else:
        p[i].get_tat(p[i-1].tat+p[i].bt)
    p[i].get_wt(p[i].tat-p[i].bt)
    PID.append(p[i].pid)
    BT.append(p[i].bt)
    WT.append(p[i].wt)
    TAT.append(p[i].tat)
    
print("\n\nSJF Scheduling")
for item in PID:
    print("< {} >".format(item),end=" ")

d={'Process #':PID,'Burst Time':BT,'Turn-around Time':TAT,'Waiting Time':WT}
ptable=pd.DataFrame(d)
display(ptable)
avw=float(sum(WT)/n)
avt=float(sum(TAT)/n)
print("Average Turnaround Time : ",avt)
print("Average Waiting Time : “,avw)

File: Scheduling/C++ Implementation/sjf.cpp
#include<iostream>

using namespace std;
struct Process{
    string pid;
    int bt;
    int wt;
    int tat;
};

bool compare(Process a , Process b){

    return a.bt <= b.bt;
}



int main()
{
    Process P[10];

    int i,n;
    float avg_wt=0,avg_tat=0;

    std::cout << "\nEnter the number of Processes";
    std::cin >> n;

    for(i=0;i<n;i++)
    {
        std::cout << "\nEnter Process ID ";
        cin >> P[i].pid;
        std::cout << "\nEnter Burst Time";
        cin >> P[i].bt;
    }

    std::sort(P,P + n,compare);

    

    

    for(i=0;i<n;i++){

        if(i==0){
            P[i].tat = P[i].bt;
        }

        else{
            P[i].tat = P[i].bt + P[i-1].tat;
        }

        P[i].wt = P[i].tat - P[i].bt;
        avg_tat += P[i].tat;
        avg_wt += P[i].wt;
    }

    std::cout <<"\nPID\t" << "BT\t" << "WT\t" << "TAT";

    for(i=0;i<n;i++){
        std::cout <<'\n'<<P[i].pid<<'\t'<<P[i].bt<<'\t'<<P[i].wt<<'\t'<<P[i].tat<<'\n';
    }

    std::cout<<"\nAverage Waiting Time : "<<avg_wt/n;
    std::cout<<"\nAverage Turnaround Time : "<<avg_tat/n<<'\n';
    
    return 0;
}
